web_generator/test/type_map_test.dart

```diff
@@ -0,0 +1,131 @@
+@TestOn('node')
+library;
+
+import 'package:test/test.dart';
+import 'package:web_generator/src/ast/builtin.dart';
+import 'package:web_generator/src/ast/declarations.dart';
+import 'package:web_generator/src/ast/types.dart';
+import 'package:web_generator/src/interop_gen/namer.dart';
+import 'package:web_generator/src/interop_gen/sub_type.dart';
+
+void main() {
+  group('Type Map Test', () {
+    test('Builtin Types Test', () {
+      final anyTypeMap = getTypeHierarchy(BuiltinType.anyType);
+      expect(anyTypeMap.nodes, isEmpty,
+          reason: 'JSAny should have no ancestors');
+
+      final booleanTypeMap =
+          getTypeHierarchy(BuiltinType.primitiveType(PrimitiveType.boolean));
+      expect(booleanTypeMap.nodes, isNotEmpty,
+          reason: 'JSBoolean inherits JSAny');
+      expect(booleanTypeMap.nodes.length, equals(1),
+          reason: 'JSBoolean only inherits JSAny');
+
+      final booleanToJSObjectLookup = booleanTypeMap.lookup('JSObject');
+      expect(booleanToJSObjectLookup, isNull,
+          reason: 'JSBoolean does not inherit JSObject');
+
+      final booleanToJSAnyLookup = booleanTypeMap.lookup('JSAny');
+      expect(booleanToJSAnyLookup, isNotNull,
+          reason: 'JSBoolean inherits JSAny');
+      final (level: booleanToJSAnyLevel, path: booleanToJSAnyPath) =
+          booleanToJSAnyLookup!;
+      expect(booleanToJSAnyLevel, equals(1));
+      expect(booleanToJSAnyPath, equals([0]));
+
+      final objectTypeMap =
+          getTypeHierarchy(BuiltinType.primitiveType(PrimitiveType.object));
+      expect(objectTypeMap.nodes.length, equals(1),
+          reason: 'JSObject only inherits JSAny');
+      final objectToObjectLookup = objectTypeMap.lookup('JSObject');
+      expect(objectToObjectLookup, isNotNull, reason: 'JSObject is JSObject');
+      final (level: objectToObjectLevel, path: objectToObjectPath) =
+          objectToObjectLookup!;
+      expect(objectToObjectLevel, equals(isZero),
+          reason: 'Lookup search on self');
+      expect(objectToObjectPath, isEmpty);
+
+      final arrayTypeMap = getTypeHierarchy(BuiltinType.primitiveType(
+          PrimitiveType.array,
+          typeParams: [BuiltinType.anyType]));
+      expect(arrayTypeMap.nodes.length, equals(1),
+          reason: 'JSArray only inherits JSObject');
+    });
+
+    test('Sub Type Primitive Test', () {
+      expect(getSubTypeOfTypes([BuiltinType.anyType, BuiltinType.anyType]).name,
+          equals('JSAny'));
+
+      final numStringSubType = getSubTypeOfTypes([
+        BuiltinType.primitiveType(PrimitiveType.num),
+        BuiltinType.primitiveType(PrimitiveType.string)
+      ]);
+      expect(numStringSubType.name, equals('JSAny'));
+    });
+
+    test('Sub Type Interface Test', () {
+      final a = InterfaceDeclaration(
```

Comment:
I'd consider a case where we check the LCA of three types where the pairwise LCAs are different than the LCA of all the types. This is to check we're not accidentally composing the LCAs (which we aren't).

e.g.

```
C -> B -> A
E -> D -> B
F -> H -> B
E -> G -> A
F -> G -> A
```

LCA{E, F} = G but LCA{C, E, F} (B) != LCA{C, LCA{E, F}} (A)